server/services/v1/auth.go

```diff
@@ -42,53 +52,194 @@ type authService struct {
+	currentSub, err := getCurrentSub(ctx)
 	if err != nil {
-		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+err.Error())
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to list applications: reason = %s", err.Error())
 	}
-	defer resp.Body.Close()
-	body, err := io.ReadAll(resp.Body)
+	currentNamespace, err := request.GetNamespace(ctx)
 	if err != nil {
-		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+err.Error())
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to list applications: reason = %s", err.Error())
 	}
-	bodyStr := string(body)
-	if resp.StatusCode == 200 {
-		getAccessTokenResponse := api.GetAccessTokenResponse{}
 
-		err = json.Unmarshal([]byte(bodyStr), &getAccessTokenResponse)
+	nonInteractiveApp := "non_interactive"
+	metadata := map[string]string{}
+	metadata[createdAt] = strconv.FormatInt(time.Now().UnixMilli(), 10)
+	metadata[createdBy] = currentSub
+	metadata[tigrisNamespace] = currentNamespace
+
+	c := &management.Client{
+		Name:           &req.Name,
+		Description:    &req.Description,
+		AppType:        &nonInteractiveApp,
+		ClientMetadata: metadata,
+	}
+
+	err = a.management.Client.Create(c)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to create application: reason = %s", err.Error())
+	}
+
+	// grant this client to access current service as audience
+	grant := &management.ClientGrant{
+		ClientID: c.ClientID,
+		Audience: &a.authConfig.Audience,
+		Scope:    []interface{}{},
+	}
+
+	err = a.management.ClientGrant.Create(grant)
+	if err != nil {
+		log.Warn().Str("app_id", c.GetClientID()).Msg("Failed to create application grant")
+		// delete this app and clean up
+		err := deleteApplication(a, c.GetClientID())
 		if err != nil {
-			return nil, api.Errorf(api.Code_INTERNAL, "Failed to parse external response: reason = "+err.Error())
+			log.Warn().Msgf("Failed to cleanup half-created app with clientId=%s, reason=%s", c.GetClientID(), err.Error())
 		}
-		return &getAccessTokenResponse, nil
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to create application grant: reason = %s", err.Error())
 	}
-	log.Error().Msgf("Auth0 response status code=%d", resp.StatusCode)
-	return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+bodyStr)
+
+	createdApp := &api.Application{
+		Name:        c.GetName(),
+		Description: c.GetDescription(),
+		Id:          c.GetClientID(),
+		Secret:      c.GetClientSecret(),
+		CreatedBy:   c.GetClientMetadata()[createdBy],
+		CreatedAt:   readCreatedAt(c),
+	}
+	return &api.CreateApplicationResponse{
+		CreatedApplication: createdApp,
+	}, nil
+}
+
+func (a *Auth0) DeleteApplication(ctx context.Context, req *api.DeleteApplicationsRequest) (*api.DeleteApplicationResponse, error) {
+	// validate
+	client, err := a.management.Client.Read(req.GetId())
```

Comment:
Is there a way to know from the error that the application doesn't exist? In that case, we don't need to return an internal error.

Reply:
updated the error handling to convey the closest status code back.